### PR TITLE
Remove werkzeug by parsing Retry-After ourselves

### LINKED ISSUE #2409: Yum install fails with RHEL v7.2
yum installation of letsencrypt.noarch and python2-letsencrypt.noarch fails when using RHEL's EPEL repository. The issue is the result of python-werkzeug - which is a dependency of both letsencrypt packages, but is not provided by EPEL. Manually installing the `python-werkzeug-0.9.1-2.el7.noarch.rpm` from the CentOS 7 Extras repository fixes the issue; installing Werkzeug by compiling from source or using pip does not resolve the issue, as yum will still list it as a failed dependency unless it sees the rpm has been installed.

Here is a list of all installed repos on the server where this fails:

Repo-id      : epel/x86_64
Repo-name    : Extra Packages for Enterprise Linux 7 - x86_64
Repo-revision: 1454890725
Repo-tags    : binary-x86_64
Repo-updated : Sun Feb  7 20:49:57 2016
Repo-pkgs    : 9,397
Repo-size    : 7.4 G
Repo-metalink: https://mirrors.fedoraproject.org/metalink?repo=epel-7&arch=x86_64
  Updated    : Sun Feb  7 20:49:57 2016
Repo-baseurl : http://s3-mirror-us-east-1.fedoraproject.org/pub/epel/7/x86_64/ (34 more)
Repo-expire  : 21,600 second(s) (last: Tue Feb  9 04:06:34 2016)
Repo-filename: /etc/yum.repos.d/epel.repo

Repo-id      : rhui-REGION-client-config-server-7/x86_64
Repo-name    : Red Hat Update Infrastructure 2.0 Client Configuration Server 7
Repo-revision: 1439271503
Repo-updated : Tue Aug 11 01:38:23 2015
Repo-pkgs    : 6
Repo-size    : 299 k
Repo-mirrors : https://rhui2-cds01.us-east-1.aws.ce.redhat.com/pulp/mirror/rhui-client-config/rhel/server/7/x86_64/os
Repo-baseurl : https://rhui2-cds01.us-east-1.aws.ce.redhat.com/pulp/repos//rhui-client-config/rhel/server/7/x86_64/os/ (1 more)
Repo-expire  : 21,600 second(s) (last: Tue Feb  9 04:06:34 2016)
Repo-filename: /etc/yum.repos.d/redhat-rhui-client-config.repo

Repo-id      : rhui-REGION-rhel-server-optional/7Server/x86_64
Repo-name    : Red Hat Enterprise Linux Server 7 Optional (RPMs)
Repo-revision: 1454007492
Repo-updated : Thu Jan 28 13:59:28 2016
Repo-pkgs    : 8,213
Repo-size    : 9.6 G
Repo-mirrors : https://rhui2-cds01.us-east-1.aws.ce.redhat.com/pulp/mirror/content/dist/rhel/rhui/server/7/7Server/x86_64/optional/os
Repo-baseurl : https://rhui2-cds02.us-east-1.aws.ce.redhat.com/pulp/repos//content/dist/rhel/rhui/server/7/7Server/x86_64/optional/os/ (1 more)
Repo-expire  : 21,600 second(s) (last: Tue Feb  9 04:06:34 2016)
Repo-filename: /etc/yum.repos.d/redhat-rhui.repo

Repo-id      : rhui-REGION-rhel-server-releases/7Server/x86_64
Repo-name    : Red Hat Enterprise Linux Server 7 (RPMs)
Repo-revision: 1454005863
Repo-updated : Thu Jan 28 13:33:14 2016
Repo-pkgs    : 10,152
Repo-size    : 12 G
Repo-mirrors : https://rhui2-cds01.us-east-1.aws.ce.redhat.com/pulp/mirror/content/dist/rhel/rhui/server/7/7Server/x86_64/os
Repo-baseurl : https://rhui2-cds01.us-east-1.aws.ce.redhat.com/pulp/repos//content/dist/rhel/rhui/server/7/7Server/x86_64/os/ (1 more)
Repo-expire  : 21,600 second(s) (last: Tue Feb  9 04:06:34 2016)
Repo-filename: /etc/yum.repos.d/redhat-rhui.repo

Repo-id      : rhui-REGION-rhel-server-rh-common/7Server/x86_64
Repo-name    : Red Hat Enterprise Linux Server 7 RH Common (RPMs)
Repo-revision: 1447968073
Repo-updated : Thu Nov 19 16:21:13 2015
Repo-pkgs    : 168
Repo-size    : 3.2 G
Repo-mirrors : https://rhui2-cds01.us-east-1.aws.ce.redhat.com/pulp/mirror/content/dist/rhel/rhui/server/7/7Server/x86_64/rh-common/os
Repo-baseurl : https://rhui2-cds02.us-east-1.aws.ce.redhat.com/pulp/repos//content/dist/rhel/rhui/server/7/7Server/x86_64/rh-common/os/ (1 more)
Repo-expire  : 21,600 second(s) (last: Tue Feb  9 04:06:34 2016)
Repo-filename: /etc/yum.repos.d/redhat-rhui.repo

### PR DESCRIPTION
**Pull request by pde**
Fixes #2409
Progress on #1301

### CODE REVIEW

**Comment by pde** [comment]
Hmmm, my scripts for rsyncing + testing were not working for `acme/`; will reopen when this is actually ready...

**Comment by pde** [comment]
May work now...

File: acme/acme/client.py [hunk 1]
```diff
@@ -245,26 +245,34 @@ def answer_challenge(self, challb, response):
 
     @classmethod
     def retry_after(cls, response, default):
-        """Compute next `poll` time based on response ``Retry-After`` header.
+        """Compute next `poll` time based on response ``Retry-After`` header,
+        per https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.37
```
Comment (kuba) [inline acme/acme/client.py]: First docstring paragraph should be limited to one line only, please move the link somewhere down.

File: acme/acme/client.py [hunk 2]
```diff
@@ -245,26 +245,34 @@ def answer_challenge(self, challb, response):
...
 
         :param requests.Response response: Response from `poll`.
         :param int default: Default value (in seconds), used when
             ``Retry-After`` header is not present or invalid.
 
         :returns: Time point when next `poll` should be performed.
-        :rtype: `datetime.datetime`
+        :rtype: `datetime.datetime.datetime`
```
Comment (kuba) [inline acme/acme/client.py]: That's incorrect

**Comment by kuba** [comment]
Rather than inlining all decoding code please write a drop-in replacement function for `werkzeug.parse_date` (with the same signature). Then write all necessary unit test cases for such function.

File: acme/acme/client.py [hunk 3]
```diff
@@ -245,26 +245,34 @@ def answer_challenge(self, challb, response):
...
             # pylint: disable=no-member
-            decoded = werkzeug.parse_date(retry_after)  # RFC1123
-            if decoded is None:
+            when = parsedate_tz(retry_after)
+            try:
+                year = when[0] # raises TypeError if t is None
+                # py26: Handle two-digit years -- but any server that thinks
+                # "retry after 99" means "come back after 1999" is.. deprecated
+                if year >= 0 and year < 100:
+                    when = [year + 2000] + list(when[1:]) #  pragma: no cover
```
Comment (kuba) [inline acme/acme/client.py]: urm, slice operator (`when[1:]`) already produces list, please don't cast explicitly

Comment (pde) [inline acme/acme/client.py]: The cast is required for python 2.6.

Comment (kuba) [inline acme/acme/client.py]: No, cast is not required for Python 2.6. The problem here is that you're trying to concatenate list and tuple which won't work with any version of Python. You should probably use `when = (year + 2000,) + when[1:]`.

File: acme/acme/client.py [hunk 4]
```diff
@@ -255,16 +258,23 @@ def retry_after(cls, response, default):
...
         """
-        retry_after = response.headers.get('Retry-After', str(default))
+        retry_after = response.headers.get('Retry-After', str(default)).strip()
         try:
             seconds = int(retry_after)
         except ValueError:
             # pylint: disable=no-member
-            decoded = werkzeug.parse_date(retry_after)  # RFC1123
-            if decoded is None:
+            when = parsedate_tz(retry_after)
```
Comment (kuba) [inline acme/acme/client.py]: According to the docs, that's using RFC 2822 (with some hacks), not RFC 2616.

File: acme/acme/client.py [hunk 5]
```diff
@@ -255,16 +258,23 @@ def retry_after(cls, response, default):
...
+            try:
+                year = when[0] # raises TypeError if t is None
+                # py26: Handle two-digit years -- but any server that thinks
+                # "retry after 99" means "come back after 1999" is.. deprecated
+                if year >= 0 and year < 100:
+                    when = [year + 2000] + list(when[1:]) #  pragma: no cover
+                tzone = when[-1] if when[-1] else 0
+                # raises ValueError/OverflowError
```
Comment (kuba) [inline acme/acme/client.py]: try/except block should be limited to the smallest possible block of code, in this case line below

File: acme/acme/client.py (hunk 3 above)
Comment (kuba) [inline acme/acme/client.py]: Also, why is this not covered? Please test it or write further inline explanation. Also, you should use `  # pragma: no cover` to satisfy linter (note double whitespace before hashtag, one after - opposite what you've done)

File: acme/acme/client.py [hunk 6]
```diff
@@ -255,16 +258,23 @@ def retry_after(cls, response, default):
         :rtype: `datetime.datetime`
 
         """
-        retry_after = response.headers.get('Retry-After', str(default))
+        retry_after = response.headers.get('Retry-After', str(default)).strip()
```
Comment (kuba) [inline acme/acme/client.py]: if you care about stripping, then please add an appropriate test case

File: acme/acme/client.py (hunk 3 above)
Comment (pde) [inline acme/acme/client.py]: It's covered in py26, but in py27 `parsedate_tz` never returns two-digit years.

File: acme/acme/client.py (hunk 4 above)
Comment (pde) [inline acme/acme/client.py]: I haven't read the RFCs closely to figure out their relationships with each other. But `werkzeug.parse_date` is a thin wrapper on `email.utils.parsedate_tz`.

Comment (pde) [inline acme/acme/client.py]: In any case, `parsedate_tz` handles [each of the 2616 cases](https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.3.1).  However I'm going to remove any special casing in our code for the RFC 850 subcase.  No ACME servers will ever speak HTTP/1.0 or issue RFC 850 dates, so we don't need extra code to sure up python 2.6's implementation of that deprecated format.

File: acme/acme/client.py (hunk 6 above)
Comment (pde) [inline acme/acme/client.py]: Done; I'm not sure if requests passes along trailing whitespace but the HTTP spec does seem to allow it.

**Comment by pde** [comment]
This should be ready for re-review.

**Comment by bmw** [comment]
@kuba, are you able to rereview this? pde really wants this in the next release. If not, I can take a crack at it but I'd much prefer you did the review if you're able/interested.

**Comment by kuba** [comment]
Yup, I'll review this. When is the next release?

BTW, why is the milestone `0.4.1`? I thought we already agreed on the versioning scheme.

File: acme/acme/client.py [hunk 7]
```diff
@@ -11,13 +11,13 @@
...
 import requests
 import sys
-import werkzeug
 
 from acme import errors
 from acme import jose
 from acme import jws
 from acme import messages
 
+from email.utils import parsedate_tz
```
Comment (kuba) [inline acme/acme/client.py]: stdlib imports should happen in the top-most import group

**Comment by kuba** [comment]
My [comment about separate drop-in function and relevant unit tests](https://github.com/letsencrypt/letsencrypt/pull/2453#issuecomment-183904010) was never fixed. I'll review this once it's fixed.

File: acme/acme/client_test.py [hunk 8]
```diff
@@ -205,6 +205,13 @@ def test_retry_after_invalid(self, dt_mock):
             datetime.datetime(2015, 3, 27, 0, 0, 10),
             self.client.retry_after(response=self.response, default=10))
 
+        # wrong date -> ValueError
```
Comment (kuba) [inline acme/acme/client_test.py]: move to separate test

File: acme/acme/client.py [hunk 9]
```diff
@@ -259,12 +262,16 @@ def retry_after(cls, response, default):
...
-            else:
-                return decoded
+            # The RFC 2822 parser handles all of RFC 2616's cases in modern
+            # environments (primarily HTTP 1.1+ but also py27+)
+            when = parsedate_tz(retry_after)
+            if when is not None:
+                try:
+                    tz_secs = datetime.timedelta(when[-1] if when[-1] else 0)
+                    return datetime.datetime(*when[:7]) - tz_secs
+                except (ValueError, OverflowError):
```
Comment (kuba) [inline acme/acme/client.py]: again, this looks like too broad `try/except`

**Comment by bmw** [comment]
@kuba, the next release is sometime this week. Preferably earlier rather than later.

And we did agree on a versioning scheme. We voted and decided to follow the semantic versioning scheme as you suggested. Initially, it was looking like this release would simply fix bugs in 0.4.0, meaning it should be called 0.4.1. If any new functionality is added, we'll call it 0.5.0.

**Comment by pde** [comment]
@kuba, werkzeug's `parse_date` function only does two things on top of `email.utils.parsedate_tz`:

1. cast the timezone from seconds into something `datetime` understands
2. play weird games -- of increasingly questionable relevance and correctness -- with two-digit years for long-deprecated RFC 822/850 support.

I don't believe we should have (2) in our tree, and I don't currently believe that (1) is enough to warrant a separate function, though I'm open to arguments on either point!

File: acme/acme/client.py (hunk 9 above)
Comment (pde) [inline acme/acme/client.py]: `OverflowError` can be raised by either the `timedelta` call or the subtract operation on the `datetime` objects.  `ValueError` can only be raised by the second line, but I believe the code as written is cleaner than two nested try blocks.

**Comment by bmw** [comment]
Again, I'd prefer not to review this and leave that to @kuba.

With that said, @pde, looking quickly at kuba's review comments, I see [this](https://github.com/letsencrypt/letsencrypt/pull/2453#issuecomment-186565114) which was never fixed if I'm understanding the comment correctly.

**Comment by pde** [comment]
@bmw my answer [here](https://github.com/letsencrypt/letsencrypt/pull/2453#issuecomment-187462269) was: I'm not sure it makes sense to have a drop-in identical replacement, but if someone has a good argument in favour I'm happy to do it.

File: acme/acme/client.py [hunk 10]
```diff
@@ -4,6 +4,7 @@
 import heapq
 import logging
 import time
+from email.utils import parsedate_tz
```
Comment (bmw) [inline acme/acme/client.py]: nit: Our inputs are alphabetized.

File: acme/setup.py [hunk 11]
```diff
@@ -20,7 +20,6 @@
     'requests',
     'setuptools',  # pkg_resources
     'six',
-    'werkzeug',
```
Comment (bmw) [inline acme/setup.py]: With this change we need to update `letsencrypt-auto-sources/pieces/letsencrypt-auto-requirements.txt` and remove this package. Otherwise, it will continue to be installed when people use `letsencrypt-auto`.

If you'd like this to be a separate PR, that's fine but you should open an issue/PR to track the issue to make sure it makes it into 0.4.1.

File: acme/acme/client.py [hunk 12]
```diff
@@ -259,12 +262,16 @@ def retry_after(cls, response, default):
...
         except ValueError:
-            # pylint: disable=no-member
-            decoded = werkzeug.parse_date(retry_after)  # RFC1123
-            if decoded is None:
-                seconds = default
-            else:
-                return decoded
+            # The RFC 2822 parser handles all of RFC 2616's cases in modern
+            # environments (primarily HTTP 1.1+ but also py27+)
+            when = parsedate_tz(retry_after)
```
Comment (bmw) [inline acme/acme/client.py]: `werkzeug` and you in a previous version of this PR call `strip` on `retry_after` ([source](https://github.com/letsencrypt/letsencrypt/blob/6fd3dba737a8de2c2f2d43d52d932418762762bb/acme/acme/client.py#L259)). I think we should be doing this unless we are **certain** it isn't necessary.

File: acme/setup.py (hunk 11 above)
Comment (pde) [inline acme/setup.py]: Thanks for spotting that!

File: acme/acme/client.py (hunk 12 above)
Comment (pde) [inline acme/acme/client.py]: I'm certain.  In py26 and py27:
```
>>> int(" 3 ")
3
>>> from email.utils import parsedate_tz
>>> parsedate_tz(" Tue, 16 Feb 2016 11:50:00 MST ")
(2016, 2, 16, 11, 50, 0, 0, 1, -1, -25200)
```

**Comment by bmw** [comment]
lgtm

**Comment by kuba** [comment]
Indeed, my concerns were never addressed, although I pointed them out at
least 3/4 times (here and on IRC). I find it ridiculous to having justify
the need of good tests or good coding practices. My time is extremely
constrained recently, so apologies for slow or sparse reviews - I prefer to
save some time by reviewing only when all changes have been implemented,
which didn't happen here. I would also prefer being pinged at least (via
Hangouts) rather than skipping LGTM. That being said, i m almost sure this
change will cause troubles for clients... If I only had time I would
request for this change to be reverted and reimplemented properly...
